libs/viewer/src/Automation.cpp

```diff
@@ -0,0 +1,262 @@
+    vector<vector<std::string>> permute;
+    for (int j = 0; j < size; ++j) {
+        const jsmntok_t tok = tokens[i];
+        CHECK_KEY(tok);
+        if (0 == compare(tok, jsonChunk, "name")) {
+            i = parse(tokens, i + 1, jsonChunk, &out->name);
+            if (VERBOSE) {
+                slog.i << "Building spec [" << out->name << "]" << io::endl;
+            }
+        } else if (0 == compare(tok, jsonChunk, "base")) {
+            i = parseBaseSettings(tokens, i + 1, jsonChunk, &base);
+        } else if (0 == compare(tok, jsonChunk, "permute")) {
+            i = parsePermutationsSpec(tokens, i + 1, jsonChunk, &permute);
+        } else {
+            slog.w << "Invalid automation key: '" << STR(tok, jsonChunk) << "'" << io::endl;
+            i = parse(tokens, i + 1);
+        }
+        if (i < 0) {
+            slog.e << "Invalid automation value: '" << STR(tok, jsonChunk) << "'" << io::endl;
+            return i;
+        }
+    }
+
+    // Determine the number of permutations.
+    size_t caseCount = 1;
+    size_t propIndex = 0;
+    vector<vector<std::string>::const_iterator> iters(permute.size());
+    for (const auto& prop : permute) {
+        caseCount *= prop.size();
+        if (VERBOSE) {
+            for (const auto& s : prop) {
+                slog.i << "  Perm: " << s.c_str() << io::endl;
+            }
+        }
+        iters[propIndex++] = prop.begin();
+    }
+    out->cases.resize(caseCount);
+
+    if (VERBOSE) {
+        slog.i << "  Case count: " << caseCount << io::endl;
+    }
+
+    size_t caseIndex = 0;
+    while (true) {
+
+        // Append a copy of the current Settings object to the case list.
+        if (VERBOSE) {
+            slog.i << "  Appending case " << caseIndex << io::endl;
+        }
+        out->cases[caseIndex++] = base;
+
+        // Leave early if there are no permutations.
+        if (iters.empty()) {
+            return i;
+        }
+
+        // Use a basic counting algorithm to generate the next test case.
+        // Bump the first digit, if it rolls back to 0 then bump the next digit, etc.
+        // In this case, the "digit" is an iterator into a vector of JSON strings.
+        propIndex = 0;
+        for (auto& iter : iters) {
+            const auto& prop = permute[propIndex++];
+            if (++iter != prop.end()) {
+                break;
+            }
+            iter = prop.begin();
+
+            // Check if all permutations have been generated.
+            if (propIndex == permute.size()) {
+                assert(caseIndex == out->cases.size());
+                return i;
+            }
+        }
+
+        // Apply changes to the settings object.
+        for (const auto& iter : iters) {
+            const std::string& jsonString = *iter;
+            if (VERBOSE) {
+                slog.i << "  Applying " << jsonString.c_str() << io::endl;
+            }
+            if (!readJson(jsonString.c_str(), jsonString.size(), &base)) {
```

Comment:
I assume after this call to `readJson`, we would do something interesting with `base` (like actually testing Filament with it). Am I understanding correctly?

Reply:
Correct, this returns a list of Settings that can then be applied to Filament; my next PR will make use of it.  The current PR merely generates the specified permutations and sets up a unit test for them.